domain/controllernode/state/state.go

```diff
@@ -146,3 +148,96 @@ WHERE  namespace = $dbNamespace.namespace`, dbNamespace)
 
 	return namespace, nil
 }
+
+// SetRunningAgentBinaryVersion sets the running agent binary version for the
+// provided controllerID. Any previously set values for this controllerID will
+// be overwritten by this call.
+//
+// The following errors can be expected:
+// - [controllernodeerrors.NotFound] if the controller node does not exist.
+// - [coreerrors.NotSupported] if the architecture is not found in the architecture database table.
+func (st *State) SetRunningAgentBinaryVersion(
+	ctx context.Context,
+	controllerID string,
+	version coreagentbinary.Version,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	arch := architecture{Name: version.Arch}
+
+	selectArchIdStmt, err := st.Prepare(`
+SELECT id AS &architecture.id FROM architecture WHERE name = $architecture.name
+`, arch)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	controllerNodeAgentVer := controllerNodeAgentVersion{
+		ControllerID: controllerID,
+		Version:      version.Number.String(),
+	}
+
+	selectControllerNodeStmt, err := st.Prepare(`
+	SELECT controller_id AS &controllerNodeAgentVersion.controller_id
+	FROM controller_node
+	WHERE controller_id = $controllerNodeAgentVersion.controller_id
+	`, controllerNodeAgentVersion{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	upsertControllerNodeAgentVerStmt, err := st.Prepare(`
+INSERT INTO controller_node_agent_version (*) VALUES ($controllerNodeAgentVersion.*)
+ON CONFLICT (controller_id) DO
+UPDATE SET version = excluded.version, architecture_id = excluded.architecture_id
+`, controllerNodeAgentVersion{})
```

Comment:
`s/controllerNodeAgentVersion{}/controllerNodeAgentVer`

Reply:
Sure thanks Ian, are we trying to ensure here that what the type sample we pass in exactly matches that of controllerNodeAgentVer?